### ISSUE TITLE
How do I get the parent of a given element?

### ISSUE THREAD

**Issue by elcinsayin** [cubewise-code/tm1py#697]
I need to get the parent of a given element. 

I checked the methods for elements and hierarchies but couldn't find a direct way to do so. I checked the hierarchy.edges but I cannot iterate through it since I don't have the parent name. The rest of the script is not quite relevant, so I feel like it is sufficient to mention that it retrieves an element name. I need to retrieve the immediate parent of that element.

Thanks in advance,

Elçin

**Comment by MariusWirtz** [comment]
Hi @elcinsayin,

we don't have a function like `tm1.elements.get_parents` in the `ElementService` to retrieve the parent of an element.
Perhaps we should implement that!

However, you can already retrieve the parent of an element through the `execute_set_mdx` function.

``` python
from TM1py import TM1Service

with TM1Service(base_url="https://localhost:12297", user="admin", password="apple") as tm1:
    mdx = "{[Week].[Week1]}"

    result_set = tm1.elements.execute_set_mdx(
        mdx=mdx,
        element_properties=["Name"],
        parent_properties=["Name"],
        member_properties=None)

    print(result_set)

```

Please bare in mind that one element can have multiple parents! In that case, the above function is not helpful, as it will always return the first parent, unless you use explicit syntax in the MDX (e.g., `{[Week].[Jan^Week1]}`).

You can also retrieve the full hierarchy and loop through the edges to find matches. Obviously, this is much less efficient on large hierarchies.

``` python
from TM1py import TM1Service
from TM1py.Utils import case_and_space_insensitive_equals

with TM1Service(base_url="https://localhost:12297", user="admin", password="apple") as tm1:
    hierarchy = tm1.hierarchies.get(dimension_name="week", hierarchy_name="week")

    for (parent, child), weight in hierarchy.edges.items():
        if case_and_space_insensitive_equals(child, "Week 1"):
            print(parent, child, weight)

```

**Comment by MariusWirtz** [comment]
The more I think about it, the more I realize we are missing a `get_parents` function in TM1py.

**Comment by macsir** [comment]
Yes, agreed. And that function should be efficient enough to get all parents by index with large number of elements. Similarly, to have a get children function?

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by elcinsayin** [comment]
Thank you so much for the answer.

**Comment by MariusWirtz** [comment]
> Yes, agreed. And that function should be efficient enough to get all parents by index with large number of elements. Similarly, to have a get children function? Get Outlook for iOS<https://aka.ms/o0ukef>

Thanks for your thoughts. 

For the get_children we actually have something in place. You can use the `get_leaves_under_consolidation` or `get_members_under_consolidation` functions.


``` python
from TM1py import TM1Service

with TM1Service(base_url="https://localhost:12297", user="admin", password="apple") as tm1:
    leaves = tm1.elements.get_leaves_under_consolidation(
        dimension_name="Week",
        hierarchy_name="Week",
        consolidation="Q1")

    print(leaves)

```

**Comment by macsir** [comment]
Thanks Marius will try it later

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
